modifications for windows build

rvuga opened the pull request:
Please let me know if you accept the changes needed in order for building the device support on windows.

ralphlange commented on opcUaDevSupApp/Makefile:
```diff
@@ -27,8 +27,13 @@ USR_INCLUDES += -I$(UASDK)/include/uaclient
 USR_INCLUDES += -I$(UASDK)/include/uapki
 USR_INCLUDES += -I$(UASDK)/include/uastack
 USR_INCLUDES += -I$(UASDK)/include/xmlparser
+USR_INCLUDES += -I$(BOOST)
```
That should be only added if BOOST is defined - else it will create an invalid commandline option.

ralphlange commented on opcUaDevSupApp/Makefile:
```diff
@@ -27,8 +27,13 @@ USR_INCLUDES += -I$(UASDK)/include/uaclient
+USR_INCLUDES += -I$(BOOST)
 
-EXTLIB_INSTALLS += $(INSTALL_LIB)/libuastack.so
+ifeq ($(OS), Windows_NT)
```
That's not the EPICS standard way check this, and might fail on other Windows type architectures.

ralphlange commented on opcUaDevSupApp/Makefile:
```diff
@@ -27,8 +27,13 @@ USR_INCLUDES += -I$(UASDK)/include/uaclient
 
-EXTLIB_INSTALLS += $(INSTALL_LIB)/libuastack.so
+ifeq ($(OS), Windows_NT)
+    EXTLIB_INSTALLS += $(INSTALL_LIB)/uastack.lib
```
Library name differences on different OSs are handled by the build system.
This might be related to #3 though.

rvuga replied:
I simplified the Makefile a bit. Unfortunately I don't know how to do it in a better way in order to solve issue #3.

ralphlange commented on opcUaDevSupApp/devOpcUa.c:
```diff
@@ -276,10 +276,13 @@ long init_longin (struct longinRecord* prec)
 
 long read_longin (struct longinRecord* prec)
 {
+    OPCUA_ItemINFO* pOPCUA_ItemINFO = NULL;
```
+1
@bkuner: While this is allowed in C++ and C99, some C compilers do not accept variable declarations in the middle of the code, only at the beginning of a block. (I.e. after any opening `{` bracket.)

ralphlange commented on opcUaDevSupApp/drvOpcUa.cpp:
```diff
@@ -801,7 +801,9 @@ long OpcWriteValue(int opcUaItemIndex,double val,int verbose)
     return 0;
 }
 /* iocShell: record write func  */
+extern "C" {
```
I don't think these are necessary.
`epicsExport.h` contains these declarations (has been for a long time).

rvuga replied:
I am afraid that the extern statements in epicsExport.h are ignored by VS compiler when expanding epicsRegisterFunction macro. Therefore they need to be repeated here otherwise we get a linking failure when building test IOC.

ralphlange replied:
Hmm. Will check how other modules handle it.

ralphlange replied:
OK, clarified.
`epicsExport.h` uses `extern "C"{}`, but only defines macros.
When you *use* the macros, those statements do not apply, and the variable you declare is getting a mangled name. GCC doesn't mangle the name anyway, so linking works.

Your solution to add the `extern "C" {}` around every use of the macro is correct.

rvuga commented:
Thank you Ralph for you comments. Please see the latest commit and my replies.

ralphlange commented:
Please bear with us for a little while longer...

I just pushed a PullRequest (#32) for issue #3 that will hopefully solve some of the issues you see on Windows.

bkuner commented:
Hello rvuga, now the program works stable there is time and I'm looking forward to Your updated pull request.

ralphlange commented:
I hope that most of the the linking / library name issues are already covered by #32.

rvuga commented:
Hi, please check the new update. The makefiles are indeed much nicer now, some additions were still needed though.

bkuner commented:
Hi rvuga, 
I checked out your branch (state 'ralphlange-fix-library-handling) and found it some patches back to my master (state R0-9). Don't know what to merge.
I thougt You want to add something that helps to compile on windows?

rvuga commented:
Hi,

can you clarify what you mean? Windows build is the aim of all of my changes. The changed files, as seen on the 'Files changed' tab are:

-README_Windows.txt:  new file
-opcUaDevSupApp/Makefile : added boost to usr_includes_win32
-opcUaDevSupApp/devOpcUa.c : changed inline call, moved declarations of some variables to starts of blocks
- opcUaDevSupApp/devUaSubscription.cpp : changed some casts
- opcUaDevSupApp/drvOpcUa.cpp : added extern C linkage to epicsRegisterFunction
- testTop/clientApp/Makefile : added win32 libraries and getopt
-  testTop/clientApp/clientMain.cpp : windows includes, and added a specific sleep call for windows
- testTop/configure/CONFIG_SITE : fixed a bug, missing '('
- testTop/testIocApp/Makefile : added win32 libraries
- testTop/testIocApp/SoftIocMain.cpp : windows includes

The way I see it these are the only differences between our branches. Did you mean something else?

ralphlange commented:
Sicher, dass Du auf den richtigen Branch schaust?

On Tue, Mar 21, 2017 at 10:43 AM, Bernhard Kuner <notifications@github.com>
wrote:

> Hi rvuga,
> I checked out your branch (state 'ralphlange-fix-library-handling) and
> found it some patches back to my master (state R0-9). Don't know what to
> merge.
> I thougt You want to add something that helps to compile on windows?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bkuner/opcUaUnifiedAutomation/pull/30#issuecomment-288025589>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAmlfy8SZ1xIi6u7cDSwtJDhvusa5Ue9ks5rn5uzgaJpZM4MBmio>
> .
>

bkuner commented:
Uups, sorry I was an the wron branch.

bkuner commented:
Uups, sorry I was an the wron branch.
